R/tag_query.R

```diff
@@ -0,0 +1,1601 @@
+# }
+
+
+# # Design notes for tagQuery:
+# ## Using stock R objects
+#
+# Advantages of standard R objects recursion
+# * Environments must be handled with care as they are pass by reference
+# * It is easy to alter the current object in place
+# * Difficult to create a new search path while altering in place
+#
+# Disadvantages of standard R objects recursion
+# * Asking for a grandparent element is difficult.
+#   * Altering a grandparent element and having the change stay is impossible
+# * Searching would need to be done at each stage, every time
+
+# ## Using environments elements
+#
+# Advantages
+# * Fast to convert to a "linked list" of tag environments
+# * Access to parents
+# * Calculations on siblings can now be done, even after alterations have been completed
+# * Once a `find(".x")` has been completed, a set of element environment pointers can be stored.
+#   * This makes followup alterations have the minimal O(k) complexity (where k is _found_ elements), not O(n) + O(k) graph search + reconstruction and k _found_ element alterations
+#
+# Disadvantages
+# * MUST be careful not alter the environment object before converting back to a list. (Ex: Do not remove the element environment's children)
+# * The item returned is a set of environments that will alter in place. We will need to be careful about documenting and/or safeguarding this
+
+
+# ## Final design choice:
+# Use environment elements
+# * Being able to search and have a list of eles to immediately look at and alter in place is AMAZING!
+# * Being able to ask for a grandparent (or obj$parent$parent) and be able to alter it in place is AMAZING! This has a strongly influenced by jquery.
+
+# ----------
+
+# # Current design decisions
+# * tagQuery objects or tag environments can NOT be used in UI. These objects MUST be converted back to standard tag objects.
+# * tagFunctions will not be altered in place
+#   * To alter tagFunction()s, use the `onRender(x)` method to register a method to be called after `as.tags(x)` is called.
+#   * `onRender(x, expr)` will wrap create a tag function that will resolve the tags before running the expr.
+
+
+## rlang::sexp_address()
+# Use to get a unique key for stacks
+# Use `env$envKey` over `rlang::sexp_address()`; 10x speed improvement
+
+# Use for `has()` functionality
+envirMap <- function() {
+  map <- fastmap()
+  list(
+    keys = map$keys,
+    asList = function() {
+      unname(map$as_list())
+    },
+    has = function(envir) {
+      map$has(envir$envKey)
+    },
+    set = function(envir, value = envir) {
+      map$set(envir$envKey, value)
+    },
+    remove = function(envir) {
+      map$remove(envir$envKey)
+    }
+  )
+}
+# Use for consistent `asList()` order
+envirStack <- function() {
+  stack <- faststack()
+  list(
+    push = stack$push,
+    asList = stack$as_list,
+    uniqueList = function() {
+      unique(stack$as_list())
+    }
+  )
+}
+
+# (Used for `unique_envirStack()` only. Do not use directly!)
```

Comment:
```suggestion
# (Used for `unique_envirStack()` only. Do not use directly!)
# Provides same interface as `envirStack()`, but checks for duplicates when
# when items are on their way in (with `push()`) instead of on the way out
# (with `uniqueList()`). This is faster when size is ~500 and above.
```